src/Lumi/Components/Form.purs

```diff
@@ -764,6 +768,59 @@ arrayModal { label, addLabel, defaultValue, summary, component, componentProps }
       , validate: pure xs
       }
 
+editableTable
+  :: forall props row result
+   . { addLabel :: String
+     , defaultValue :: row
+     , form :: FormBuilder { readonly :: Boolean | props } row result
+     , labels :: Array String
+     , maxRows :: Int
+     , summary :: JSX
+     }
+  -> FormBuilder
+      { readonly :: Boolean
+      | props
+      }
+      (Array row)
+      (Array result)
+editableTable { addLabel, defaultValue, form, labels, maxRows, summary } =
+  FormBuilder \props@{ readonly } rows ->
+    let
+      fs = (un FormBuilder form) props <$> rows
+    in
+      { edit: \onChange -> pure $
+          Child
+            { key: Nothing
+            , child:
+                EditableTable.editableTable
+                  { addLabel
+                  , maxRows
+                  , readonly
+                  , rows: Left (Array.mapWithIndex Tuple rows)
+                  , rowEq: unsafeRefEq
+                  , summary
+                  , onRowAdd: onChange (flip Array.snoc defaultValue)
+                  , onRowRemove: \(Tuple i _) -> onChange \rows' -> fromMaybe rows' (Array.deleteAt i rows')
+                  , columns: labels # Array.mapWithIndex \j label ->
+                      { label
+                      , renderCell: \(Tuple i _) -> fold do
+                          { edit } <- Array.index fs i
+                          tree <- Array.index (edit \f -> onChange (\rows' -> fromMaybe rows' (Array.modifyAt i f rows'))) j
+                          -- TODO: improve this.
+                          let naiveRender = case _ of
+                                Child { child } ->
+                                  child
+                                Wrapper { children } ->
+                                  foldMap naiveRender children
+                                Node { children, validationError } ->
+                                  foldMap naiveRender children
+                          pure $ naiveRender tree
```

Comment:
What improvement do you have in mind?

Reply:
Something configurable via props?

Reply:
Actually I was thinking about something in the types that would allow us to restrict the type of UI representation that these forms could have, as `Forest` is way too general to fit inside a table cell.

Reply:
This ought to be using the keys when matching on the `Tree` constructors presumably?